### DISCUSSION TITLE
Gesture Events and V6

### DISCUSSION THREAD

**Discussion by milonic** [fabricjs/fabric.js#9733] [Q&A]
Hi Guys,

How are things going with touch events and V6?

events.js (which served us well) definitely isn't the best solution anymore. Although a decent bit of code in its day, it's bloated with obsolete references to MS stuff, etc. Plus, it doesn't look like it's supported anymore and hasn't been for some time.

I don't think hammerjs is the way to go either.

I think touch:start/end etc are also obsolete as we can do everything with simple Pointer events now and browser support for Pointer events is good at 97%.

I have created my own gesture code working in V6 - it can pan, pinch zoom, rotate, longpress and multi hit using one or two fingers. Only problem is: I had to expose the pointer events within fabric to get it to work. Mapping them to mouse events with enablePointerEvents would not let me detect the second finger. Could be the way fabric modifies the events, I didn't look into it in much detail.

Anyway, please let me know where we are at as I'd like to get gestures moving.

**Comment by ShaMan123** [comment]
I have seen an impl as you describe that had to override some methods.
The author said he can PR but I can't say if or when.

**Comment by ShaMan123** [comment]
I can contibute a rotate and scale gesture

**Comment by asturur** [comment]
I have been using hammer.js and i was ok with that to the point i was wondering if a support for gesture is needed at all.
There are some interesting points, one of them is that i don't see standard solutions in the wild, the most modern i found is hammer.js and is abandoned as well.
There are many gestures you may want to have, but if you go outside:
- 2 finger pan vs drag
- rotate
- scale
Then is your own ux territory. ( if you have an example of common gesture please add, maybe i miss some )

So my first question is why i can't find a good event library to abstract gestures.

The second question i have is that gestures once identified becomes higher level event on top of pointer down for example. Once you have an implementation for it, what kind of support fabricJS should even offer?

The gesture happens on the canvas and is an event, like a click.
Now fabricJS decided to support interaction and so we have a long trail of code after a click happens, that start a selection or a transform.
But for a gesture?
On top of informing the developer  that a user is doing a pinch action, what else would you do?
Deciding if the user is:
- zooming the canvas,
- scaling an object
- doing a special selection action

is entirely up to the developer decision and i don't think we should go in that businness.

**Comment by asturur** [comment]
> Only problem is: I had to expose the pointer events within fabric to get it to work. Mapping them to mouse events with enablePointerEvents would not let me detect the second finger. Could be the way fabric modifies the events, I didn't look into it in much detail.

Exposing the pointer events and stop mixing them with mouse events could be something that makes developer more free to build their solution.
But also i wonder why you had to hook up to the canvas events. Couldn't you just hook up to the html element event handlers?

If you want to share a single implementation for the purpose of a discussion do it

**Comment by milonic** [comment]
> But also i wonder why you had to hook up to the canvas events. Couldn't you just hook up to the html element event handlers?

I never even gave that a thought. Let me have a play and see what happens.

This is a great discussion. The whole concept of how to implement gestures is not an easy one, I've been toying with it for a while now and still cannot decide what would be best. 

Although I think I might be leaning towards a modular approach though as opposed to hard coding it into fabricjs. 

If I can just get mutiple down/up pointer events handled that might be all that is needed.

**Reply by ShaMan123** [reply]
> If I can just get mutiple down/up pointer events handled that might be all that is needed.

I agree

**Reply by asturur** [reply]
Imho we need to work in a direction that allow the developer to wrap the canvas with it, for example:

```
import { addGestures, } from 'fabric/accessories';

addGestures(myCanvasInstance);
```

Possibly this would attach to html elements, and provide something back that is specific of `addGestures` and should be unopinionated, meaning that can provide handlers for pinch but shouldn't attach them but should make the canvas and objects fire a pinch object.

This is just an example.

**Reply by ShaMan123** [reply]
I would argue that all interactions should be like that, plugins, not default behavior

**Comment by milonic** [comment]
Here's my input from my last few hours on this.

So, I moved my events from within fabric to parent HTML elements. I tried window, document.body and a canvas container div. All seemed to work OK so I began some more testing.

Anyway, I have had to go back to inserting them as fabric events because it broke the context menu that is part my system. Not 100% sure yet but it looks like pointer events are overriding the mouse and touch events that I use for the context menu. I'm surprised as I assumed all three (mouse, touch and pointer) events would fire simultaneously but something went wrong.

So, in my VERY humble opinion I suggest looking at a way of incorporating native pointer events within fabric, it just seems tidier that way.

I did look into how you could get the none primary event to register but it's not trivial. The isMainEvent() function is blocking it. When I removed the function I could get both events to register but it was buggy. Messing with isMainEvent is probably a bad idea.

I wonder if there's any way you could get native pointer events exposed in Fabric but stick with a plugin for the gestures? I think that might work. 

Maybe have a exposePointerEvents parameter as well as enablePointerEvents for compatability? Just a thought.

**Reply by asturur** [reply]
isMainEvent is there just to try to recognize the main click from accidental multi touch events, since fabric isn't multi touch doing anything.
If we decide to implement a gesture plugin 'isMainEvent()' needs to be removed or redefined, probably the pointer/mouse duality needs to be handled differently. 

But it isn't really helpful to have native events proxyed by fabric imho.
To have a gesture functionality you need the full scope of the event in the fabric event, at that point, why do i have to pass it to you?  In order to pass it to you i have to add the html element listener to the upper canvas (or window ) at that pint the gesture plugin can just do that.

It is also possible that without having a working implementation to consult right now i miss some important part.

Also, again, with HammerJS 2 years ago everything was working smoothly and the integration points with fabricJS were very few and no overrides, so i m sure it can be done without too much overhaul. Anyway i ll add this item to the roadmap for a documentation/feasibility check so someone can look at it deeper.

**Comment by asturur** [comment]
@milonic check this discussion here: https://github.com/fabricjs/fabric.js/discussions/9736
If you feel like you can try to do something similar for gesture support so that it can find its place in a roadmap.
If you feel strongly about an implementation pattern or another try to describe it, then we look forward on how to do that with less as a principle. At least from my side. If you come with an idea that let you implement gestures with 2 lines of code i ll ask you if we can get to 1 of course.
I have other features i want to make the proposal for so i won't get to it sooner, if you want to accelerate it do it.
Reserve the second and third message after your description so the important things are all in the upper part of the page

**Comment by 07ke** [comment]
Has anyone implemented, pinch to zoom, pinch to rotate? I've been stumped for a long time with no way to get started. 🥺

**Reply by Smrtnyk** [reply]
you can use hammerjs for that
this is how I did it https://github.com/Smrtnyk/Firetable/blob/master/packages/floor-creator/src/TouchManager.ts#L27

**Reply by molexx** [reply]
Thanks @Smrtnyk but that source doesn't seem to be available any more and my google-fu can't find it, would it be possible to copy the relevant section to a gist or something please?
